src/dir-hamt-sharded.js

```diff
@@ -27,41 +32,180 @@ function shardedDirExporter (cid, node, name, path, pathRest, resolve, size, dag
+  waterfall([
+    (cb) => {
+      if (!options.rootBucket) {
+        options.rootBucket = new Bucket({
+          hashFn: DirSharded.hashFn
+        })
+        options.hamtDepth = 1
+
+        return addLinksToHamtBucket(node.links, options.rootBucket, options.rootBucket, cb)
+      }
+
+      return addLinksToHamtBucket(node.links, options.lastBucket, options.rootBucket, cb)
+    },
+    (cb) => findPosition(targetFile, options.rootBucket, cb),
+    ({ position }, cb) => {
+      let prefix = toPrefix(position.pos)
+      const bucketPath = toBucketPath(position)
+
+      if (bucketPath.length > (options.hamtDepth)) {
+        options.lastBucket = bucketPath[options.hamtDepth]
+
+        prefix = toPrefix(options.lastBucket._posAtParent)
+      }
+
+      const streams = [
+        pull(
+          values(node.links),
+          map((link) => {
+            const entryPrefix = link.name.substring(0, 2)
+            const entryName = link.name.substring(2)
+            const entryPath = entryName ? path + '/' + entryName : path
+
+            if (entryPrefix !== prefix) {
+              // not the entry or subshard we're looking for
+              return false
+            }
+
+            if (entryName && entryName !== targetFile) {
+              // not the entry we're looking for
+              return false
+            }
+
+            if (!entryName) {
+              // we are doing to descend into a subshard
+              options.hamtDepth++
+            } else {
+              // we've found the node we are looking for, remove the context
+              // so we don't affect further hamt traversals
+              delete options.rootBucket
+              delete options.lastBucket
+              delete options.hamtDepth
+            }
+
+            return {
+              depth: entryName ? depth + 1 : depth,
+              name: entryName,
+              path: entryPath,
+              multihash: link.cid.buffer,
+              pathRest: entryName ? pathRest.slice(1) : pathRest,
+              parent: dir || parent
+            }
+          }),
+          filter(Boolean),
+          resolve
+        )
+      ]
+
+      if (options.fullPath) {
+        streams.unshift(values([dir]))
+      }
+
+      cb(null, streams)
+    }
+  ], (err, streams) => {
+    if (err) {
+      return deferred.resolve(error(err))
+    }
+
+    deferred.resolve(cat(streams))
+  })
+
+  return deferred
+}
+
+const addLinksToHamtBucket = (links, bucket, rootBucket, callback) => {
+  Promise.all(
+    links.map(link => {
+      if (link.name.length === 2) {
+        const pos = parseInt(link.name, 16)
+
+        return bucket._putObjectAt(pos, new Bucket({
+          hashFn: DirSharded.hashFn
+        }, bucket, pos))
+      }
+
+      return rootBucket.put(link.name.substring(2), true)
+    })
+  )
+    .catch(err => {
+      callback(err)
+      callback = null
+    })
+    .then(() => callback && callback())
```

Comment:
If `.then` is invoked and `callback` throws, `catch` is invoked which calls `callback`.  If `callback` is from `async.js` it'll be wrapped in `once` which throws another error saying `callback already invoked!` and smothering the actual error.

All this will go away when we start using async/await/iterators everywhere.

Reply:
Ahh, gotcha